android/views/src/main/kotlin/com/boswelja/ephemeris/views/CalendarPagerAdapter.kt

```diff
@@ -49,50 +48,106 @@ internal class CalendarPagerAdapter : InfinitePagerAdapter<CalendarPageViewHolde
 
     override fun onBindHolder(holder: CalendarPageViewHolder, page: Int) {
         val pageState = pageLoader!!.getPageData(page)
-        holder.bindDisplayRows(dayBinder!!, pageState)
+        holder.bindDisplayRows(pageLoader!!, dayBinder!!, pageState)
     }
 }
 
 internal class CalendarPageViewHolder(
     private val binding: CalendarPageBinding
-) : RecyclerView.ViewHolder(binding.root) {
+) : ViewHolder(binding.root) {
 
     private val inflater = LayoutInflater.from(itemView.context)
 
+    private val rowBindingCache = mutableListOf<CalendarRowBinding>()
+    private val dateCellViewHolderCache = mutableListOf<ViewHolder>()
+
+    private var boundDateCells = 0
+    private var trimDateCellCache = false
+
+    private var dateBinder: CalendarDateBinder<ViewHolder>? = null
+        set(value) {
+            if (field != value) {
+                field = value
+                // Clear our cached date cells on binder change
+                dateCellViewHolderCache.clear()
+            }
+        }
+    private var pageLoader: CalendarPageLoader? = null
+        set(value) {
+            if (field != value) {
+                field = value
+                // Only invalidate row cache. date cells are trimmed after binding to optimize CPU
+                rowBindingCache.removeAll {
```

Comment:
I'm tempted to say 
```kotlin
rowBindingCache.onEach { 
    it.root.removeAllViewsInLayout() 
}.clear()
```
Will be better here. We don't need the predicate to decide if we're always clearing the whole list, so there's no extra decisions to make on each iteration

Reply:
I figured `clear` iterates the whole list again, but if that's not the case then `onEach` absolutely makes more sense

Reply:
ah yes it does, 
```java
// clear to let GC do its work
for (int i = 0; i < size; i++)
    elementData[i] = null;

size = 0;
```
love it, just nulls out each element, sets the list size to 0 and lets GC clear it :joy:

Reply:
Yep in that case iterating once should be quicker, would be nice if we didn't need the predicate but I don't see an existing extension for it sadly